### ISSUE TITLE
Use absolute domain names for endpoints

### ISSUE THREAD

**Issue by jml** [boto/botocore#1767]
https://github.com/boto/botocore/blob/develop/botocore/data/endpoints.json specifies relative domain names for endpoints. For example, `api.ecr.us-west-1.amazonaws.com`. 

On many Kubernetes deployments, looking up a relative DNS name can lead to many sequential DNS requests, which affects end-user latency (https://pracucci.com/kubernetes-dns-resolution-ndots-options-and-why-it-may-affect-application-performances.html is a great summary). 

We can work around this locally by configuring `ndots`, but if botocore used absolute DNS names, e.g. `api.ecr.us-west-1.amazonaws.com.` (note final `.`), this would not be an issue for us, at least for AWS requests. 

I was initially going to submit a PR for this, but noticed that endpoints.json is regularly updated by an automated tool which I couldn't find the source code for.

**Comment by jamesls** [comment]
The `endpoints.json` file is autogenerated from an upstream source that we automatically pull in.  This file (and the data in the file) is shared by most of the AWS SDKs and other tools, so any changes to this file could potentially impact a large number of people, so we'd need to really vet this change.

I'm not really familiar with the specific kubernetes issue here, but reading the article you linked it says:

> It’s not a definitive solution but can be a quick and dirty change to apply to improve the situation. This is the “hot patch” we applied to resolve the issue, whose results have been shown in the screenshots above.

I'm a little hesitant to make this change for a file that's consumed by a lot of libraries/tools based on that.  Do you have any more info on this?

**Comment by jml** [comment]
I'm not sure what to say.

Adding a trailing `.` to a relative domain name turns it into an absolute
domain name. This is part of the DNS RFC (sorry, I'm offline and don't have
link to hand). I'm not in a position to assess how widely supported it is
among DNS servers, but I'd imagine it's close to universal.

Without the trailing `.`, any client will do one DNS look up per item in
the search path (generally configured in /etc/resolv.conf). Most systems
don't have many things in the search path, so this isn't a big deal, but
it's still a performance hit.

There's an optimisation you can use which is that if a DNS name has enough
`.`s in it, the lookup will be done on the absolute domain _first_ and on
search paths second. It's a heuristic built on the idea that almost no one
wants to look up `ec2.us-west1.amazonaws.com.default.svc.cluster.local.`.

I get that this is a risky change, and given that there are workarounds, I
think it would be reasonable to just not do it. On the other hand, if
you've got a way of managing the risk, it has the potential to make a whole
bunch of clients magically faster.

**Comment by joguSD** [comment]
Given the risk and effort required to make this change, I think we're going to pass on this feature request for now. I tried forcing this through our event system and ran into issues with SigV4 signing. It might still be possible if I tried a little more but I don't have much more time to investigate how you'd implement an event handler for this.

Events documentation for reference:
https://boto3.amazonaws.com/v1/documentation/api/latest/guide/events.html
https://botocore.amazonaws.com/v1/documentation/api/latest/topics/events.html
